vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1706 @@
+#include <kj/std/iostream.h>
+#include <zlib.h>
+#include <algorithm>
+#include <map>
+#include <queue>
+
+#include "rr_graph_fpga_interchange.h"
+
+#include "arch_util.h"
+#include "vtr_time.h"
+#include "vtr_error.h"
+#include "globals.h"
+#include "check_rr_graph.h"
+#include "rr_metadata.h"
+#include "rr_graph_indexed_data.h"
+#include "fpga_interchange_arch_utils.h"
+
+//#define DEBUG
+
+using namespace DeviceResources;
+using namespace LogicalNetlist;
+using namespace capnp;
+
+static const auto INOUT = LogicalNetlist::Netlist::Direction::INOUT;
+
+typedef std::pair<int, int> location;
+
+enum node_sides {
+    LEFT_EDGE = 0,
+    TOP_EDGE,
+    RIGHT_EDGE,
+    BOTTOM_EDGE,
+};
+
+auto NODESIDES = {LEFT_EDGE, TOP_EDGE, RIGHT_EDGE, BOTTOM_EDGE};
+enum node_sides OPPOSITE_SIDE[] = {RIGHT_EDGE, BOTTOM_EDGE, LEFT_EDGE, TOP_EDGE};
+
+/*
+ * Intermediate data type.
+ * It is used to build and explore FPGA Interchange
+ * nodes graph and later convert them to rr_nodes.
+ */
+struct intermediate_node {
+  public:
+    location loc;
+    int segment_id;
+    std::vector<bool> links; //left, up, right, down
+    intermediate_node* visited;
+    bool back_bone;
+    bool on_route;
+    bool has_pins;
+    intermediate_node() = delete;
+    intermediate_node(int x, int y)
+        : loc{x, y} {
+        links.resize(4);
+        for (auto i : NODESIDES)
+            links[i] = false;
+        visited = nullptr;
+        on_route = false;
+        has_pins = false;
+        back_bone = false;
+        segment_id = -1;
+    }
+    intermediate_node(location loc_)
+        : loc(loc_) {
+        links.resize(4);
+        for (auto i : NODESIDES)
+            links[i] = false;
+        visited = nullptr;
+        on_route = false;
+        has_pins = false;
+        back_bone = false;
+        segment_id = -1;
+    }
+};
+
+/*
+ * Main class for FPGA Interchange device to rr_graph flow
+ */
+struct RR_Graph_Builder {
+  public:
+    RR_Graph_Builder(Device::Reader& arch_reader,
+                     const DeviceGrid& grid,
+                     DeviceContext& device_ctx,
+                     const std::vector<t_physical_tile_type>& physical_tile_types,
+                     const vtr::vector<RRSegmentId, t_segment_inf>& segment_inf,
+                     const enum e_base_cost_type base_cost_type)
+        : ar_(arch_reader)
+        , device_ctx_(device_ctx)
+        , segment_inf_(segment_inf)
+        , grid_(grid)
+        , base_cost_type_(base_cost_type)
+        , empty_(device_ctx.arch->strings.intern_string(vtr::string_view(""))) {
```

Comment:
What is this used for?

Reply:
segment_inf_ is used in multiple places in code. mainly in packing intermediate data to final rr_graph.
grid_ is used to create mapping from virtual channels to physical and is required for rr_graph_builder.
base_cost_type_ is only used in final stages of rr_graph generation to run all_and_load_rr_indexed_data
empty_ is indeed unused in rest of the code, it'll be removed in next commit